### DISCUSSION TITLE
onBeforeRender() + pinia

### DISCUSSION THREAD

**Discussion by hamedg68** [vikejs/vike#993] [Discussion]
### Description

**onBeforeRender()** does not work with pinia, actually I want to use **onBerforeRender()** for both server and client side in **.page.ts** file, I got the following error while pinia is initialized.
there is a way which solve the problem, using **onServerPrefetch()** and **onMounted()** hook  according to pinia example, but I want to use **onBeforeRender()**

### Error Stack

```shell
Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?
```

**Comment by brillout** [comment]
~~Please respect https://brillout.github.io/rules/.~~ @hamedg68 Do you still have issues? I'm happy to assist.

**Comment by apappas1129** [comment]
I am confused with the difference between `onBeforeRender` and Vue's `onServerPrefetch`. Can someone help me with when I should use which? I am using Client Routing, `enableEagerStreaming`, and `pinia`.

**Reply by brillout** [reply]
`onBeforeRender()` is global: it's run before the *entire* page is rendered. Whereas it seems (I'm not all too familiar with Vue) that `onServerPrefetch()` is run before *the component* is rendered.

I'd be inclined to say: if you want to fetch data in a global fashion, i.e. for your entire page, then `onBeforeRender()` is probably a good choice. If you want to collocate your data fetching logic next to your component, then `onBeforeRender()` can't be used but it may be possible with Vue's `onServerPrefetch()`.
